### PR TITLE
MT-25 Remove CRTP and protection violations from Worker based classes

### PR DESCRIPTION
**Pull request by wiesiekpap**
This solves the issue of some virtual calls out of the base destructor, clash between different shutdown/destruction paths, unclear access with friend classes plus a number of small bugs. The mutual exclusion of shutdown actions addresses SEGFAULT observed in testing.
Also, some spurious additional delays and unneeded thread yields have been removed, together with a couple of the most obnoxious of the many const_casts.

### CODE REVIEW

File: src/blockchain/node/base/SyncServer.hpp [hunk 1]
```diff
@@ -113,11 +113,24 @@ class SyncServer : public SyncDM, public SyncWorker
         ::zmq_connect(socket_.get(), endpoint_.c_str());
     }
 
-    ~SyncServer() { signal_shutdown().get(); }
+    ~SyncServer() final
+    {
+        try {
+            Shutdown().get();
+        } catch (const std::exception&) {
+            // TODO MT-34 improve
```
Comment (bartalke) [inline src/blockchain/node/base/SyncServer.hpp:122]: I suggest at least print what went wrong, before MT-34 will be implemented

File: src/blockchain/node/base/SyncServer.hpp [hunk 2]
```diff
@@ -484,4 +448,61 @@ class SyncServer : public SyncDM, public SyncWorker
...
+auto SyncServer::pipeline(zmq::Message&& in) noexcept -> void
+{
+    if (false == running_.load()) { return; }
+
+    const auto body = in.Body();
+
+    OT_ASSERT(1 <= body.size());
+
+    using Work = Work;
+    const auto work = [&] {
```
Comment (bartalke) [inline src/blockchain/node/base/SyncServer.hpp:462]: It seems that only &body can be captured

File: src/blockchain/node/blockoracle/BlockDownloader.cpp [hunk 3]
```diff
@@ -237,5 +242,12 @@ auto BlockDownloader::update_tip(const Position& position, const int&)
     socket_->Send(std::move(work));
 }
 
-BlockDownloader::~BlockDownloader() { signal_shutdown().get(); }
+BlockDownloader::~BlockDownloader()
+{
+    try {
+        Shutdown().get();
+    } catch (const std::exception&) {
+        // TODO MT-34 improve
```
Comment (bartalke) [inline src/blockchain/node/blockoracle/BlockDownloader.cpp:248]: I suggest at least print what went wrong, before MT-34 will be implemented

File: src/blockchain/node/filteroracle/BlockIndexer.cpp [hunk 4]
```diff
@@ -398,5 +402,12 @@ auto FilterOracle::BlockIndexer::update_tip(
     notify_(type_, position);
 }
 
-FilterOracle::BlockIndexer::~BlockIndexer() { signal_shutdown().get(); }
+FilterOracle::BlockIndexer::~BlockIndexer()
+{
+    try {
+        signal_shutdown().get();
+    } catch (const std::exception&) {
+        // TODO MT-34 improve
```
Comment (bartalke) [inline src/blockchain/node/filteroracle/BlockIndexer.cpp:408]: I suggest at least print what went wrong, before MT-34 will be implemented

File: src/blockchain/node/filteroracle/FilterDownloader.hpp [hunk 5]
```diff
@@ -95,11 +96,24 @@ class FilterOracle::FilterDownloader : public FilterDM, public FilterWorker
         init_executor({shutdown});
     }
 
-    ~FilterDownloader() { signal_shutdown().get(); }
+    ~FilterDownloader() final
+    {
+        try {
+            signal_shutdown().get();
+        } catch (const std::exception&) {
+            // TODO MT-34 improve
```
Comment (bartalke) [inline src/blockchain/node/filteroracle/FilterDownloader.hpp:105]: I suggest at least print what went wrong, before MT-34 will be implemented

File: src/blockchain/node/filteroracle/FilterDownloader.hpp [hunk 6]
```diff
@@ -222,12 +198,60 @@ class FilterOracle::FilterDownloader : public FilterDM, public FilterWorker
...
+    -> void
+{
+    if (false == running_.load()) { return; }
+
+    const auto body = in.Body();
+
+    OT_ASSERT(0 < body.size());
+
+    using Work = FilterOracle::Work;
+    const auto work = [&] {
```
Comment (bartalke) [inline src/blockchain/node/filteroracle/FilterDownloader.hpp:214]: As previous seems that only &body should be captured.

File: src/blockchain/node/filteroracle/HeaderDownloader.hpp [hunk 7]
```diff
@@ -77,11 +78,24 @@ class FilterOracle::HeaderDownloader : public HeaderDM, public HeaderWorker
         OT_ASSERT(checkpoint_);
     }
 
-    ~HeaderDownloader() { signal_shutdown().get(); }
+    ~HeaderDownloader() final
+    {
+        try {
+            signal_shutdown().get();
+        } catch (const std::exception&) {
+            // TODO MT-34 improve
```
Comment (bartalke) [inline src/blockchain/node/filteroracle/HeaderDownloader.hpp:87]: I suggest at least print what went wrong, before MT-34 will be implemented

File: src/blockchain/node/filteroracle/HeaderDownloader.hpp [hunk 8]
```diff
@@ -252,12 +224,64 @@ class FilterOracle::HeaderDownloader : public HeaderDM, public HeaderWorker
...
+auto FilterOracle::HeaderDownloader::pipeline(zmq::Message&& in) noexcept
+    -> void
+{
+    if (false == running_.load()) { return; }
+
+    const auto body = in.Body();
+
+    OT_ASSERT(1 <= body.size());
+
+    const auto work = [&] {
```
Comment (bartalke) [inline src/blockchain/node/filteroracle/HeaderDownloader.hpp:239]: Seems that only &body should be captured

File: src/blockchain/p2p/peer/Address.cpp [hunk 9]
```diff
@@ -7,6 +7,7 @@
 #include "1_Internal.hpp"                   // IWYU pragma: associated
 #include "blockchain/p2p/peer/Address.hpp"  // IWYU pragma: associated
 
+#include <iostream>
```
Comment (bartalke) [inline src/blockchain/p2p/peer/Address.cpp]: I guess it is not needed ?

File: src/blockchain/node/base/Base.cpp [hunk 10]
```diff
@@ -1186,25 +1180,12 @@ auto Base::SendToPaymentCode(
     return SendToPaymentCode(nymID, recipient.asBase58(), amount, memo);
 }
 
-auto Base::shutdown(std::promise<void>& promise) noexcept -> void
+auto Base::shut_down() noexcept -> void
```
Comment (radexx99) [inline src/blockchain/node/base/Base.cpp:1183]: why such name with undescore ? it oveerides sth different with such name ?

File: src/blockchain/node/filteroracle/BlockIndexer.hpp [hunk 11]
```diff
@@ -122,11 +122,17 @@ class FilterOracle::BlockIndexer : public BlockDMFilter,
         const UnallocatedCString& shutdown,
         const NotifyCallback& notify) noexcept;
 
-    ~BlockIndexer();
+    ~BlockIndexer() final;
+
+protected:
+    auto pipeline(zmq::Message&& in) noexcept -> void final;
```
Comment (radexx99) [inline src/blockchain/node/filteroracle/BlockIndexer.hpp:128]: What is the added value of changing definition of this pipeline ? performance ?

File: src/blockchain/p2p/peer/Address.cpp [hunk 12]
```diff
@@ -39,8 +42,7 @@ auto Address::Chain() const noexcept -> blockchain::Type
 auto Address::Display() const noexcept -> UnallocatedCString
 {
     Lock lock(lock_);
-
-    return address_->Display();
+    return address_ ? address_->Display() : std::string("BOTCHED");
```
Comment (radexx99) [inline src/blockchain/p2p/peer/Address.cpp]: What is Botched ? maybe we can add some static member to provide this string value ?

File: src/interface/ui/base/List.hpp [hunk 13]
```diff
@@ -94,7 +94,7 @@ class List : virtual public ExternalInterface,
     {
         return qt_model_;
     }
-    auto shutdown(std::promise<void>& promise) noexcept -> void
+    auto xshutdown(std::promise<void>& promise) noexcept -> void
```
Comment (radexx99) [inline src/interface/ui/base/List.hpp]: why this change ? to remove inheritance ? why 'x' ? does it mean sth or it is just random letter ?

**Review by radexx99 — APPROVED** [review]

File: src/interface/ui/activitythread/ActivityThread.cpp [hunk 14]
```diff
@@ -985,6 +992,10 @@ auto ActivityThread::validate_account(
 ActivityThread::~ActivityThread()
 {
     wait_for_startup();
-    signal_shutdown().get();
+    try {
+        signal_shutdown().get();
+    } catch (const std::exception&) {
+        // TODO MT-34 improve
```
Comment (bartalke) [inline src/interface/ui/activitythread/ActivityThread.cpp:996]: I suggest to print what went wrong before MT-34 will be implemented

File: src/interface/ui/base/List.hpp (hunk 13 above)
Comment (bartalke) [inline src/interface/ui/base/List.hpp]: what is xshutdown ?

File: src/interface/ui/base/List.hpp [hunk 15]
```diff
@@ -226,7 +225,14 @@ class List : virtual public ExternalInterface,
 
         return std::future_status::ready == startup_future_.wait_for(none);
     }
-    auto wait_for_startup() const noexcept -> void { startup_future_.get(); }
+    auto wait_for_startup() const noexcept -> void
+    {
+        try {
+            startup_future_.get();
+        } catch (const std::exception&) {
+            // TODO MT-34 improve
```
Comment (bartalke) [inline src/interface/ui/base/List.hpp:225]: I suggest to print what went wrong before MT-34 will be implemented

File: src/interface/ui/blockchainaccountstatus/BlockchainAccountStatus.cpp [hunk 16]
```diff
@@ -507,5 +516,16 @@ auto BlockchainAccountStatus::subchain_display_name(
...
+BlockchainAccountStatus::~BlockchainAccountStatus()
+{
+    try {
+        if (check_running_mark_stopped()) {
+            close_pipeline();
+            // TODO MT-34 investigate what other actions might be needed
+            mark_shut_down();
+        }
+    } catch (const std::exception&) {
+        // TODO MT-34 diagnostics
```
Comment (bartalke) [inline src/interface/ui/blockchainaccountstatus/BlockchainAccountStatus.cpp]: I suggest to print what went wrong before MT-34 will be implemented

File: src/blockchain/p2p/peer/Address.cpp (hunk 12 above)
Comment (polishcode) [reply src/blockchain/p2p/peer/Address.cpp]: Second remark - BOTCHED path should notify us of an erroneous condition; method is noexcept, but is returning that string strong enough to report the problem?

File: src/blockchain/node/base/Base.cpp (hunk 10 above)
Comment (wiesiekpap) [reply src/blockchain/node/base/Base.cpp:1183]: 'shutdown' is used a lot in other cases. Besides, I always try to use function names that include a verb, that makes the code a bit more readable.

File: src/blockchain/node/base/SyncServer.hpp (hunk 2 above)
Comment (wiesiekpap) [reply src/blockchain/node/base/SyncServer.hpp:462]: Yes, but this lambda will be removed, so we are not investing in it.

File: src/blockchain/node/filteroracle/BlockIndexer.hpp (hunk 11 above)
Comment (wiesiekpap) [reply src/blockchain/node/filteroracle/BlockIndexer.hpp:128]: No. Originally, some of the subclasses had Message passed as rvalue reference, others as const reference, with the static downcast silently adjusting the call. With a virtual pipeline that now avoids the downcast, all the overrides have to have the same signature.

File: src/blockchain/p2p/peer/Address.cpp (hunk 12 above)
Comment (wiesiekpap) [reply src/blockchain/p2p/peer/Address.cpp]: I'll remove this testing leftover. When the problem occurred, it was massive, this was just the firs place it struck when running one particular test. It is one of the cases where the problem just have to be prevented rather than detected.

File: src/interface/ui/base/List.hpp (hunk 13 above)
Comment (wiesiekpap) [reply src/interface/ui/base/List.hpp]: It is a proof that once the subclasses each have their own shut_down, no-one needs this shutdown() any more, well spotted.

File: src/blockchain/p2p/peer/Address.cpp (hunk 9 above)
Comment (wiesiekpap) [reply src/blockchain/p2p/peer/Address.cpp]: Right, done.

File: src/blockchain/node/blockoracle/BlockDownloader.cpp (hunk 3 above)
Comment (wiesiekpap) [reply src/blockchain/node/blockoracle/BlockDownloader.cpp:248]: More is needed, hence the TODO and ticket, but I have added a minimal diagnostic line especially for you :-)

File: src/blockchain/node/base/SyncServer.hpp (hunk 1 above)
Comment (wiesiekpap) [reply src/blockchain/node/base/SyncServer.hpp:122]: More is needed, hence the TODO and ticket, but I have added a minimal diagnostic line especially for you :-)

File: src/blockchain/node/filteroracle/BlockIndexer.cpp (hunk 4 above)
Comment (wiesiekpap) [reply src/blockchain/node/filteroracle/BlockIndexer.cpp:408]: More is needed, hence the TODO and ticket, but I have added a minimal diagnostic line especially for you :-)

File: src/blockchain/node/filteroracle/FilterDownloader.hpp (hunk 5 above)
Comment (wiesiekpap) [reply src/blockchain/node/filteroracle/FilterDownloader.hpp:105]: More is needed, hence the TODO and ticket, but I have added a minimal diagnostic line especially for you :-)

File: src/blockchain/node/filteroracle/FilterDownloader.hpp (hunk 6 above)
Comment (wiesiekpap) [reply src/blockchain/node/filteroracle/FilterDownloader.hpp:214]: As previously mentioned, that is the least of the problems, it will have to go and it is beyond the scope of this stage.

File: src/blockchain/node/filteroracle/HeaderDownloader.hpp (hunk 7 above)
Comment (wiesiekpap) [reply src/blockchain/node/filteroracle/HeaderDownloader.hpp:87]: More is needed, hence the TODO and ticket, but I have added a minimal diagnostic line especially for you :-)

File: src/blockchain/node/filteroracle/HeaderDownloader.hpp (hunk 8 above)
Comment (wiesiekpap) [reply src/blockchain/node/filteroracle/HeaderDownloader.hpp:239]: It will disappear, but in a next stage.

File: src/interface/ui/activitythread/ActivityThread.cpp (hunk 14 above)
Comment (wiesiekpap) [reply src/interface/ui/activitythread/ActivityThread.cpp:996]: More is needed, hence the TODO and ticket, but I have added a minimal diagnostic line especially for you :-)

File: src/interface/ui/base/List.hpp (hunk 15 above)
Comment (wiesiekpap) [reply src/interface/ui/base/List.hpp:225]: More is needed, hence the TODO and ticket, but I have added a minimal diagnostic line especially for you :-)

File: src/interface/ui/blockchainaccountstatus/BlockchainAccountStatus.cpp (hunk 16 above)
Comment (wiesiekpap) [reply src/interface/ui/blockchainaccountstatus/BlockchainAccountStatus.cpp]: More is needed, hence the TODO and ticket, but I have added a minimal diagnostic line especially for you :-)

**Review by adamtkaczyk — APPROVED** [review]

**Review by bartalke — APPROVED** [review]

File: src/blockchain/database/common/Peers.cpp [hunk 17]
```diff
@@ -202,29 +202,29 @@ auto Peers::Import(UnallocatedVector<Address_p> peers) noexcept -> bool
     return insert(lock, std::move(newPeers));
 }
 
-auto Peers::Insert(Address_p pAddress) noexcept -> bool
+auto Peers::Insert(Address_p&& pAddress) noexcept -> bool
 {
     auto peers = UnallocatedVector<Address_p>{};
-    peers.emplace_back(std::move(pAddress));
+    peers.push_back(std::move(pAddress));
```
Comment (TomaszAugustyn) [inline src/blockchain/database/common/Peers.cpp:208]: This can be simplified with:
`auto peers = UnallocatedVector<Address_p>{ std::move(pAddress) };`

File: src/blockchain/node/base/SyncServer.hpp [hunk 18]
```diff
@@ -484,4 +449,61 @@ class SyncServer : public SyncDM, public SyncWorker
         ::zmq_disconnect(socket_.get(), endpoint_.c_str());
     }
 };
+
+auto SyncServer::pipeline(zmq::Message&& in) noexcept -> void
+{
+    if (false == running_.load()) { return; }
```
Comment (TomaszAugustyn) [inline src/blockchain/node/base/SyncServer.hpp:455]: Maybe just `if (!running_.load())` ? There are also other places like this.

File: src/blockchain/node/filteroracle/FilterOracle.hpp [hunk 19]
```diff
@@ -195,7 +195,6 @@ class FilterOracle final : virtual public node::internal::FilterOracle
     ~FilterOracle() final;
 
 private:
-    friend Worker<FilterOracle, api::Session>;
     friend internal::FilterOracle;
```
Comment (TomaszAugustyn) [inline src/blockchain/node/filteroracle/FilterOracle.hpp:198]: Are we also going to get rid of the rest of the friends in future?

File: src/blockchain/node/filteroracle/HeaderDownloader.hpp [hunk 20]
```diff
@@ -252,12 +225,64 @@ class FilterOracle::HeaderDownloader : public HeaderDM, public HeaderWorker
...
+};
+
+auto FilterOracle::HeaderDownloader::pipeline(zmq::Message&& in) noexcept
+    -> void
+{
+    if (false == running_.load()) { return; }
+
+    const auto body = in.Body();
+
+    OT_ASSERT(1 <= body.size());
```
Comment (TomaszAugustyn) [inline src/blockchain/node/filteroracle/HeaderDownloader.hpp:237]: Why not like in other places: `OT_ASSERT(0 < body.size());` ?

File: src/blockchain/p2p/Address.hpp [hunk 21]
```diff
@@ -96,9 +96,10 @@ class Address final : public internal::Address
         const Time lastConnected,
         const UnallocatedSet<Service>& services,
         const bool incoming) noexcept(false);
-    Address(const Address& rhs) noexcept;
 
-    ~Address() final = default;
+    ~Address() final;
+
+    Address(const Address* rhs) noexcept;
```
Comment (TomaszAugustyn) [inline src/blockchain/p2p/Address.hpp:102]: One argument ctors tend to be declared `explicit` to avoid implicit conversions.

File: src/core/Worker.hpp [hunk 22]
```diff
@@ -87,59 +102,72 @@ class Worker
...
         LogTrace()(OT_PRETTY_CLASS())("using ZMQ batch ")(pipeline_.BatchID())
             .Flush();
     }
 
-    ~Worker() { stop_worker().get(); }
+    virtual ~Worker()
+    {
+        if (check_running_mark_stopped()) {
+            close_pipeline();
+            mark_shut_down();
```
Comment (TomaszAugustyn) [inline src/core/Worker.hpp]: Even though `mark_shut_down` is marked `noexcept`, `std::call_once` used there can still throw `std::system_error` if any condition prevents calls to `std::call_once` from executing as specified. And as we know throwing in dectructors is not a good idea. There are a lot of places where `mark_shut_down` is called in destructor.

File: src/blockchain/node/filteroracle/FilterOracle.hpp (hunk 19 above)
Comment (wiesiekpap) [reply src/blockchain/node/filteroracle/FilterOracle.hpp:198]: Generally yes, possibly allowing a few exceptions for testing and external operators - I'm not a great fan of these exceptions, but neither do I want to appear dogmatic and infexible :-)

File: src/blockchain/p2p/Address.hpp (hunk 21 above)
Comment (wiesiekpap) [reply src/blockchain/p2p/Address.hpp:102]: Agreed. This whole Address business is so dirty I'll need to think a bit about its future. The present limited changes were just to help chase a segfault that was generated when an instance of Address was being prematurely deleted.

File: src/core/Worker.hpp (hunk 22 above)
Comment (wiesiekpap) [reply src/core/Worker.hpp]: You are right. The whole business of how noexcept qualifiers and const qualifiers are used in this code is a great source of concern. We'll need to address the issue as part of the refactor.
